substrate/primitives/core/src/lib.rs

```diff
@@ -197,6 +197,36 @@ pub trait TypeId {
 	const TYPE_ID: [u8; 4];
 }
 
+/// A log level as it is represented in the runtime.
+///
+/// Used for marshalling log levels through the FFI boundary.
+#[derive(Encode, Decode, Copy, Clone)]
+pub enum RuntimeInterfaceLogLevel {
```

Comment:
Why do we need a second version of this? Can we not just fix the `LogLevel` discriminants?

Reply:
How? The old runtimes already incorporate the now-existing version. For them, `Error` is 0. And for the old node software, `Error` is 1. If we want backward compatibility, we need to convert.

Implementing `TryFrom` returning a different value is still possible, but, as @koute has put it, it'd be weird for `log_level as u8` to give a different result than `u8::from(log_level)`, and I do agree.

Doing that conversion explicitly adds some amount of code, but code transparency and readability are worth it, I believe.

Reply:
> And for the old node software, `Error` is 1. If we want backward compatibility, we need to convert.

The important piece here is the FFI encoding, which is having `Error` at `0`. And that would also work with all the old nodes.

Reply:
> so we shouldn't care about maintaining backward compatibility when passing it back?

I don't get what you mean by this? We need to ensure that the encoding start at `0` and nothing else. The node otherwise doesn't use these discriminants. (Something to check, but we can probably find out by removing Encode/Decode)

Reply:
I mean, it's a public type, and thus a public interface. In pure theory, someone might use it to pass a value in the opposite direction, from the host to the runtime, and the change you propose breaks it. I understand that in practice, that is very unlikely to happen, and anyone using (or better said abusing) the interfaces in such a way is digging their own grave. Still, I couldn't avoid bringing it up as a concern (so if the theoretical users described appear at some point, I could point them to this comment).

I still think the double conversion is a cleaner solution, but let's discuss it in the morning with a fresh mind. It's usually a bad idea to make tactical code changes at night :sweat_smile:

Reply:
> to me that whoever wrote it meant for these discriminants

HAHAHA. I bet with you that this is a verbatim copy of the enum from the log crate at this time. Don't ask me why this was done back then, I don't see any reason for it in the original pr.

I really don't see for all the fluff around this type. I also have high doubts that anyone externally uses this. Especially the discriminants. However, I think it would be more reasonable to move this type closer to the actual runtime interface. Maybe give it a different name. This way people are aware that the stuff changed and if they depended on this, their code will break.

Reply:
Okay, so the plan is:
* Remove `LogLevel` entirely and only leave `RuntimeInterfaceLogLevel`. Use implicit discriminants (as it is what `.enumerate` in the old `PassByEnum` did in practice)
* Remove `ConvertAndPassAs` introduced by the current version of PR and pass the types normally via `PassAs`
* Migrate `StorageKind` and `HttpError` exactly the same way (renaming the types and using implicit discriminants)
* Hope that nothing explodes.

Any corrections?

Reply:
You can also use explicit discriminant, just need to use the "correct' ones :)  Otherwise it sounds good.